File: addons/comm_whatsapp_chatbot/controllers/agent_workspace.py
# -*- coding: utf-8 -*-
"""Agent Workspace endpoints — backs the OWL client action that agents use
during live voice calls.

Endpoints (all type='json', auth='user' — only authenticated agents reach them):
    /voice/start        Create a call session against a chatbot + contact
    /voice/turn         Advance the engine (agent typed / clicked something)
    /voice/update       Apply slot edits without advancing
    /voice/end          Close the session with an outcome + wrap-up notes
    /voice/setup        Return the chatbot's slot definitions for the right pane
"""

import json
import logging

from odoo import http
from odoo.http import request

_logger = logging.getLogger(__name__)


class AgentWorkspaceController(http.Controller):

    @http.route("/voice/start", type="json", auth="user", methods=["POST"], csrf=False)
    def start(self, chatbot_id=None, contact_details=None, **_kw):
        """Create or reuse a contact + open a new call session."""
        if not chatbot_id:
            return {"error": "Missing chatbot_id"}
        try:
            details = contact_details or {}
            mobile = (details.get('mobile') or '').strip()
            name = (details.get('name') or '').strip()
            if not mobile:
                return {"error": "Customer mobile is required to start a call."}

            Partner = request.env['res.partner'].sudo()
            partner = Partner.search([('mobile', '=', mobile)], limit=1)
            if not partner:
                partner = Partner.create({
                    'name': name or f"Caller {mobile}",
                    'mobile': mobile,
                    'is_company': False,
                })
            elif name and partner.name != name:
                # Don't relabel existing real partners — just leave their name as-is.
                pass

            Contact = request.env['whatsapp.chatbot.contact'].sudo()
            contact = Contact.search([('partner_id', '=', partner.id)], limit=1)
            if not contact:
                contact = Contact.create({
                    'partner_id': partner.id,
                })

            Session = request.env['comm.voice.call.session'].sudo()
            session = Session.create({
                'chatbot_id': int(chatbot_id),
                'contact_id': contact.id,
                'agent_id': request.env.user.id,
            })
            return {
                "session_id": session.id,
                "contact_id": contact.id,
                "partner": {
                    "id": partner.id, "name": partner.name, "mobile": partner.mobile,
                    "email": partner.email or '',
                },
            }
        except Exception as e:
            _logger.error("Workspace start failed: %s", e, exc_info=True)
            return {"error": str(e)}

    @http.route("/voice/turn", type="json", auth="user", methods=["POST"], csrf=False)
    def turn(self, session_id=None, user_input=None, initial_variables=None, **_kw):
        if not session_id:
            return {"error": "Missing session_id"}
        try:
            return request.env["whatsapp.chatbot.message"].sudo().agent_turn(
                call_session_id=int(session_id),
                user_input=user_input,
                initial_variables=initial_variables,
            )
        except Exception as e:
            _logger.error("Workspace turn failed: %s", e, exc_info=True)
            return {"error": str(e)}

    @http.route("/voice/update", type="json", auth="user", methods=["POST"], csrf=False)
    def update(self, session_id=None, initial_variables=None, **_kw):
        if not session_id:
            return {"error": "Missing session_id"}
        try:
            return request.env["whatsapp.chatbot.message"].sudo().agent_update_state(
                call_session_id=int(session_id),
                initial_variables=initial_variables,
            )
        except Exception as e:
            _logger.error("Workspace update failed: %s", e, exc_info=True)
            return {"error": str(e)}

    @http.route("/voice/end", type="json", auth="user", methods=["POST"], csrf=False)
    def end(self, session_id=None, outcome='resolved', notes=None, **_kw):
        if not session_id:
            return {"error": "Missing session_id"}
        try:
            sess = request.env["comm.voice.call.session"].sudo().browse(int(session_id)).exists()
            if not sess:
                return {"error": "Session not found"}
            sess.action_close(outcome=outcome, notes=notes)
            return {"ok": True, "outcome": sess.outcome, "duration": sess.duration_seconds}
        except Exception as e:
            _logger.error("Workspace end failed: %s", e, exc_info=True)
            return {"error": str(e)}

    @http.route("/voice/setup", type="json", auth="user", methods=["POST"], csrf=False)
    def setup(self, chatbot_id=None, contact_id=None, **_kw):
        """Return slot definitions (variables) for the chatbot + linked bots,
        with the contact's saved values prefilled. Mirrors the simulator's
        /chatbot/simulate/setup so the right-pane dashboard can render."""
        if not chatbot_id:
            return {"bots": []}
        try:
            data = request.env["whatsapp.chatbot.message"].sudo().simulator_setup(
                chatbot_id=int(chatbot_id),
                contact_details={},  # we look up by contact below
            )
            if contact_id:
                contact = request.env["whatsapp.chatbot.contact"].sudo().browse(int(contact_id)).exists()
                if contact:
                    saved = {v.variable_id.id: (v.value or '')
                             for v in contact.variable_value_ids
                             if v.variable_id}
                    for bot in data.get('bots', []):
                        for v in bot.get('variables', []):
                            v['value'] = saved.get(v['id'], v.get('value', ''))
                    # also expose pivot_text per variable so the workspace can
                    # render it as a tooltip hint.
                    Variable = request.env['whatsapp.chatbot.variable'].sudo()
                    for bot in data.get('bots', []):
                        for v in bot.get('variables', []):
                            rec = Variable.browse(v['id']).exists()
                            if rec:
                                v['pivot_text'] = rec.pivot_text or ''
            return data
        except Exception as e:
            _logger.error("Workspace setup failed: %s", e, exc_info=True)
            return {"bots": []}

File: addons/comm_whatsapp_chatbot/static/src/js/agent_workspace.js
/** @odoo-module **/

import { Component, markup, onMounted, useState } from "@odoo/owl";
import { registry } from "@web/core/registry";
import { useService } from "@web/core/utils/hooks";
import { rpc } from "@web/core/network/rpc";

/* ── Markdown / escaping helpers (mirror the flow widget's) ─────────────── */
function bodyToHtml(text) {
    if (!text) return "";
    let s = String(text)
        .replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
    s = s.replace(/```([^`\n]+)```/g, "<code>$1</code>");
    s = s.replace(/\*([^*\n]+)\*/g, "<strong>$1</strong>");
    s = s.replace(/_([^_\n]+)_/g, "<em>$1</em>");
    s = s.replace(/~([^~\n]+)~/g, "<s>$1</s>");
    return s.replace(/\n/g, "<br>");
}
function plainToHtml(text) {
    if (!text) return "";
    return String(text)
        .replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;")
        .replace(/\n/g, "<br>");
}

export class AgentWorkspace extends Component {
    static template = "comm_whatsapp_chatbot.AgentWorkspace";
    static props = ["action", "actionId?"];

    setup() {
        this.notification = useService("notification");
        this.action = useService("action");

        this.chatbotId = this.props.action.params?.chatbot_id;
        this.chatbotName = this.props.action.params?.chatbot_name || "";

        this.state = useState({
            // Pre-call form
            started: false,
            personaName: "",
            personaMobile: "",
            // Live session
            sessionId: null,
            partner: null,
            bubbles: [],
            userInput: "",
            sending: false,
            terminate: false,
            // Slot dashboard
            setupBots: [],
            setupLoading: false,
            // Wrap-up
            showWrap: false,
            outcome: "resolved",
            notes: "",
            saving: false,
            // WhatsApp call integration — soft dependency on the
            // comm_whatsapp_calling module.
            waCallEnabled: !!(this.env.services && this.env.services.comm_whatsapp_calling),
            waCallDial: true,      // pre-call toggle
            waCallActive: false,   // set true once dial() has been fired
        });

        onMounted(() => this._loadSetup());
    }

    _wa() {
        // Lazy access to the calling service so we don't blow up when the
        // optional module isn't installed.
        return this.env.services?.comm_whatsapp_calling || null;
    }

    /* ── Setup ──────────────────────────────────────────────────────────── */
    async _loadSetup() {
        this.state.setupLoading = true;
        try {
            const data = await rpc("/voice/setup", {
                chatbot_id: this.chatbotId,
                contact_id: null,
            });
            this.state.setupBots = data?.bots || [];
        } finally {
            this.state.setupLoading = false;
        }
    }

    /* ── Pre-call ───────────────────────────────────────────────────────── */
    async _startCall() {
        const mobile = (this.state.personaMobile || "").trim();
        if (!mobile) {
            this.notification.add("Customer mobile is required.", { type: "warning" });
            return;
        }
        this.state.sending = true;
        try {
            const data = await rpc("/voice/start", {
                chatbot_id: this.chatbotId,
                contact_details: {
                    name: (this.state.personaName || "").trim(),
                    mobile,
                },
            });
            if (data?.error) {
                this.notification.add(data.error, { type: "danger" });
                this.state.sending = false;
                return;
            }
            this.state.sessionId = data.session_id;
            this.state.partner = data.partner;
            this.state.started = true;

            // Fire the real WhatsApp call in parallel when the toggle is
            // on and the calling service is available. Any dial failure
            // stays soft — the coaching session still runs.
            if (this.state.waCallEnabled && this.state.waCallDial) {
                const svc = this._wa();
                if (svc) {
                    try {
                        this.state.waCallActive = true;
                        svc.dialCall({
                            toNumber:    mobile,
                            partnerId:   data.partner?.id || null,
                            partnerName: data.partner?.name || mobile,
                            chatbotId:   this.chatbotId || null,
                        }).catch((err) => {
                            this.state.waCallActive = false;
                            this.notification.add(
                                "WhatsApp dial failed: " + (err?.message || err),
                                { type: "warning" });
                        });
                    } catch (err) {
                        this.state.waCallActive = false;
                    }
                }
            }
            // Re-load setup now that we have a contact (so saved values prefill).
            await this._loadSetupForContact(data.contact_id);
            // First engine turn — produces the welcome script.
            await this._sendTurn(null);
        } catch (e) {
            this.notification.add("Could not start the call.", { type: "danger" });
            this.state.sending = false;
        }
    }

    async _loadSetupForContact(contactId) {
        try {
            const data = await rpc("/voice/setup", {
                chatbot_id: this.chatbotId,
                contact_id: contactId,
            });
            this.state.setupBots = data?.bots || [];
        } catch (e) { /* non-fatal */ }
    }

    /* ── Turn driving ───────────────────────────────────────────────────── */
    _collectVariables() {
        const out = [];
        for (const bot of this.state.setupBots) {
            for (const v of (bot.variables || [])) {
                if (v.value !== "" && v.value !== null && v.value !== undefined) {
                    out.push({ variable_id: v.id, value: v.value });
                }
            }
        }
        return out;
    }

    async _sendTurn(userInput) {
        this.state.sending = true;
        try {
            const data = await rpc("/voice/turn", {
                session_id: this.state.sessionId,
                user_input: userInput,
                initial_variables: this._collectVariables(),
            });
            if (data?.error) {
                this.notification.add(data.error, { type: "danger" });
                return;
            }
            for (const b of (data.bubbles || [])) {
                this.state.bubbles.push({
                    ...b,
                    dir: "in",
                    text: b.text || b.body || "",
                    step_type: b.step_type || "message",
                });
            }
            this.state.terminate = !!data.terminate;
            // Refresh slot values from server (engine may have set some).
            await this._refreshSlots();
        } catch (e) {
            this.notification.add("Engine error.", { type: "danger" });
        } finally {
            this.state.sending = false;
            queueMicrotask(() => {
                const el = document.querySelector(".o_aw_script");
                if (el) el.scrollTop = el.scrollHeight;
            });
        }
    }

    async _refreshSlots() {
        // Re-fetch setup so the right-pane reflects engine-set values.
        if (!this.state.sessionId) return;
        try {
            const data = await rpc("/voice/setup", {
                chatbot_id: this.chatbotId,
                contact_id: (await this._getContactIdFromSession()),
            });
            // Preserve any local-only edits the agent is typing right now: we
            // only update slots whose current displayed value is empty. (Cheap
            // protection against clobbering work-in-progress text.)
            const newBots = data?.bots || [];
            for (const newBot of newBots) {
                const existing = this.state.setupBots.find(b => b.chatbot_id === newBot.chatbot_id);
                if (!existing) continue;
                for (const newVar of (newBot.variables || [])) {
                    const cur = existing.variables.find(v => v.id === newVar.id);
                    if (cur && (!cur.value || cur.value === '')) {
                        cur.value = newVar.value;
                    }
                }
            }
        } catch (e) { /* non-fatal */ }
    }

    async _getContactIdFromSession() {
        // We stored partner.id at start; the contact id wasn't echoed back to
        // the client originally. The setup endpoint accepts the contact_id
        // directly — to keep this simple in v1, we just pass null and accept
        // that prefill happens only at start time. The /voice/turn loop
        // already keeps the engine state consistent.
        return null;
    }

    _sendTyped() {
        const txt = (this.state.userInput || "").trim();
        if (!txt || this.state.sending || this.state.terminate) return;
        this.state.bubbles.push({ text: txt, dir: "out", step_type: "user" });
        this.state.userInput = "";
        this._sendTurn(txt);
    }

    _sendQuickReply(text) {
        if (!text || this.state.sending || this.state.terminate) return;
        this.state.bubbles.push({ text, dir: "out", step_type: "user" });
        this._sendTurn(text);
    }

    _onInputKey(ev) {
        if (ev.key === "Enter") {
            ev.preventDefault();
            this._sendTyped();
        }
    }

    /* ── Slot edits ─────────────────────────────────────────────────────── */
    async _applySlotEdits() {
        this.state.sending = true;
        try {
            await rpc("/voice/update", {
                session_id: this.state.sessionId,
                initial_variables: this._collectVariables(),
            });
            this.notification.add("Slot values saved.", { type: "success" });
        } catch (e) {
            this.notification.add("Could not save slot values.", { type: "danger" });
        } finally {
            this.state.sending = false;
        }
    }

    /* ── Navigation ─────────────────────────────────────────────────────── */
    _navigateToChatbotForm() {
        // act_window_close is for modal dialogs — it doesn't pop a client
        // action off the stack. Explicit doAction → chatbot form view works
        // reliably regardless of how the workspace was opened.
        this.action.doAction({
            type:      "ir.actions.act_window",
            res_model: "whatsapp.chatbot",
            res_id:    this.chatbotId,
            views:     [[false, "form"]],
            target:    "current",
        });
    }

    _goBack() {
        // Workspace closes; the session stays open server-side. Agents
        // can resume it from Voice Call Sessions or just walk away. Hint
        // appears only if the session is still in flight.
        if (this.state.started && !this.state.terminate) {
            this.notification.add(
                "Workspace closed. Session is still open — resume from Voice Call Sessions.",
                { type: "info" }
            );
        }
        this._navigateToChatbotForm();
    }

    /* ── Wrap-up ────────────────────────────────────────────────────────── */
    _openWrap() { this.state.showWrap = true; }
    _closeWrap() { this.state.showWrap = false; }

    async _endCall() {
        if (!this.state.sessionId) return;
        this.state.saving = true;
        // Hang up the live WhatsApp call first so the customer's phone
        // stops ringing while the coaching session closes.
        if (this.state.waCallActive) {
            const svc = this._wa();
            try { svc?.hangupActive?.(); } catch (e) { /* noop */ }
            this.state.waCallActive = false;
        }
        try {
            await rpc("/voice/end", {
                session_id: this.state.sessionId,
                outcome: this.state.outcome,
                notes: this.state.notes,
            });
            this.notification.add("Call closed.", { type: "success" });
            // Close the wrap-up overlay AND navigate back. Setting showWrap
            // false first means even if navigation is slow the user doesn't
            // see the dialog hanging around.
            this.state.showWrap = false;
            this._navigateToChatbotForm();
        } catch (e) {
            this.notification.add("Could not close the call.", { type: "danger" });
        } finally {
            this.state.saving = false;
        }
    }

    /* ── Renderers exposed to the template ─────────────────────────────── */
    formatBody(text) { return markup(bodyToHtml(text)); }
    escape(text) { return markup(plainToHtml(text)); }
}

registry.category("actions").add("comm_whatsapp_chatbot.agent_workspace", AgentWorkspace);
